### PR TITLE
Equal operator fix

### LINKED ISSUE #1003: Data Prepper Expression Evaluator
**Is your feature request related to a problem? Please describe.**
- [ ] Evaluate parsed data prepper expressions 
- [ ] Handle parsing and evaluate exceptions
- [ ] Use Dependency Injection to instantiate evaluation classes
- [ ] Add e2e test for parsing -> evaluation

**Describe the solution you'd like**

## UML Diagram
when-processor-Updated UML (5)

## Sequence Diagram
when-processor-Control Flow (2)

### _interface_ ExpressionEvaluator

Primary interface of the `org.opensearch.dataprepper.expression` package. Exposes functionality expression evaluation API to other packages. All other interfaces and classes are package private.

### _class_ ConditionalExpressionEvaluator

Default implementation of the `ExpressionEvaluator` interface. Controls the flow of execution in the following order

1. Parse the statement using the `Parser` service
2. Evaluates the statement using the `Evaluator` service
3. Cource the result using the `CourcionService`
4. Return a boolean representing the evaluated results

### _interface_ Parser

Describes `Parser` functionality.

### _class_ ParseTreeParser

Default implementation of the `Parser` interface. Converts a statement String to a `ParseTree` object using ANTLR generated classes.

### _interface_ Evaluator

Describes `Evaluator` functionality.

### _class_ ParseTreeEvaluator

Default implementation of the `Evaluator` interface. Uses and ANTLR `ParseTreeWalker` to travers a `ParseTree`. The `ParseTreeWalker` will call corresponding functions on `ParseTreeEvaluatorListener` instance on entering or exiting a node and on visiting a leaf (Terminal Node).

### _class_ ParseTreeEvaluationListener

Implements `DataPrepperExpressionListener` (extends `ParseTreeEvaluatorListener` and generated by ANTLR). Handles all tree traversal events using an operand stack and a argument stack to track visited nodes. On exit rule events, both stacks are queried to determine if an operation strategy should be triggered. All available operation strategies are registed when the ParseTreeEvaluationListener is constructed.

### _class_ CourcionService

Service to contain object type detection, argument count verification and type conversion (casting).

### _interface_ Operation

Interface every operation must implement to be registered as an evaluation strategy for the `ParseTreeEvaluationListener`.

**Describe alternatives you've considered (Optional)**
n/a

**Additional context**
Related to RFC #522 
Depends on #1001

### PR DESCRIPTION
**Pull request by sbayer55**
### Description
Added support for equality checks with different numeric types or null values
 
### Issues Resolved
#1003 
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java [hunk 1]
```diff
@@ -8,41 +8,41 @@
...
+    private final Map<Class<?>, Map<Class<?>, BiFunction<Object, Object, Boolean>>> equalStrategy;
 
-    public GenericEqualOperator(final int symbol, BiPredicate<Object, Object> operation) {
-        this.symbol = symbol;
-        displayName = DataPrepperExpressionParser.VOCABULARY.getDisplayName(symbol);
-        this.operation = operation;
+    public GenericEqualOperator(
+            final int symbol,
+            final int shouldEvaluateRuleIndex,
+            final Map<Class<?>, Map<Class<?>, BiFunction<Object, Object, Boolean>>> equalStrategy
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java:22]: This seems a very rigid strategy. As the client of this code you must make a very complicated map.

Why not just use a `BiFunction<Object, Object, Boolean>` as the strategy? Then let the caller make a more complicated use of `Map` if necessary for that particular situation?

Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java:22]: What advantage does `BiFunction` bring over `BiPredicate`?

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        );
+    }
+
+    private static Stream<Arguments> validTestArgProvider() {
+        return Stream.of(
+                Arguments.of(10, 10, true),
+                Arguments.of(10, 10.0f, true),
+                Arguments.of(-10.0f, -10, true),
+                Arguments.of(2000000000, 2000000000, true),
+                Arguments.of(2000000000, 2000000000.1f, true),
```
Comment (dlvenable) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java:41]: This seems to be missing a test case of comparing two float values.

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        assertThat(
+                lhs + " == " + rhs + " should equal is " + isArgsEqual,
+                genericEqualOperator.evaluate(lhs, rhs),
+                is(isArgsEqual)
+        );
+    }
+
+    private static Stream<Arguments> validTestArgProvider() {
+        return Stream.of(
+                Arguments.of(10, 10, true),
```
Comment (dlvenable) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java:37]: We should have some negative test cases for some of these same combinations.

e.g. `Arguments.of(10, 9, false)`.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java [hunk 4]
```diff
@@ -110,12 +118,35 @@ public GenericRegexMatchOperator regexNotEqualOperator() {
 
     @Bean
     public GenericEqualOperator equalOperator() {
-        return new GenericEqualOperator(DataPrepperExpressionParser.EQUAL, equals);
+        final BiFunction<Object, Object, Boolean> floatEquals = (lhs, rhs) -> ((Number)lhs).floatValue() == ((Number) rhs).floatValue();
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java:121]: I'm not sure I see the value of having all this code here and not in the equals class. As the number of cases expands, we'd need to cram even more into this function.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.stream.Stream;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.isA;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@ExtendWith(MockitoExtension.class)
+class OperatorConfigurationTest {
```
Comment (chenqi0805) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java:19]: Can we merge the test cases here into EqualOperatorTest and NotEqualOperatorTest?

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java:17]: It seems that all the `GenericEqualOperator` should care about is whether two objects are equal. Thus, the only strategy it needs is `BiPredicate<Object, Object>`. (I have another comment asking predicate vs. function below).

So this code should be able to run:

```
return operation.test(leftOperand, rightOperand);
```

The caller will determine how this `BiPredicate` works. In the current case, the caller does it by making the map. We could even have a common class to help with this.

```
public class ClassBasedBiFunction implements BiFunction<Object, Object> {
  private final Map<Class<?>, Map<Class<?>, BiPredicate<Object, Object>>> lhsToRhsToPredicate;
  private final BiPredicate<Object, Object> defaultOperation;
  public ClassBasedBiFunction(final Map<Class<?>, Map<Class<?>, BiPredicate<Object, Object>>> lhsToRhsToPredicate, BiPredicate<Object, Object> defaultOperation) {
  this. lhsToRhsToPredicate = lhsToRhsToPredicate;
  }

  public boolean test(Object lhs, Object rhs) {
    if (hasStrategyMatching(lhs, rhs)) {
             return equalStrategy.get(lhs.getClass())
                     .get(rhs.getClass())
                     .apply(lhs, rhs);
         }
         else {
             return defaultOperation.test(lhs, rhs);
         }
  }

  private boolean hasStrategyMatching(final Object leftOperand, final Object rightOperand) {
    // more or less the same
  }
}

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java [hunk 6]
```diff
@@ -8,41 +8,41 @@
...
+import java.util.Map;
+import java.util.function.BiFunction;
 
 import static com.google.common.base.Preconditions.checkArgument;
 
-public class GenericEqualOperator implements Operator<Boolean> {
-    private final int symbol;
-    private final String displayName;
-    private final BiPredicate<Object, Object> operation;
+public class GenericEqualOperator extends BinaryOperator<Boolean> {
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java:11]: Does this class (or its constructor) need to be public?

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericNotOperator.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package org.opensearch.dataprepper.expression;
+
+import org.antlr.v4.runtime.RuleContext;
+
+public class GenericNotOperator extends BinaryOperator<Boolean> {
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericNotOperator.java:8]: Can this be package protected?

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java:17]: I see we are already using this pattern:

https://github.com/opensearch-project/data-prepper/blob/a37d557cf64740a98b53af2fd40a8324d8cec807/data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/NumericCompareOperator.java#L21-L22

So I'm fine with continuing it.

Comment (sbayer55) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java:17]: Great suggestions on the BiPredicate, I was previously unfamiliar with that interface.

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java (hunk 5 above)
Comment (sbayer55) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/OperatorConfigurationTest.java:24]: I scoped EqualOperatorTest and NotEqualOperatorTest to cover only the functionality of the corresponding class. GenericEqualOperator itself is not concerned with the result of comparing specific values until it is configured.

In my opinion testing the configured options here is better reflection of the separation of responsibility.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java [hunk 8]
```diff
@@ -17,7 +17,7 @@
 import java.util.function.BiPredicate;
 
 @Named
-public class OperatorFactory {
+public class OperatorConfiguration {
```
Comment (chenqi0805) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java:20]: Nit: we can make it package private

**Review by chenqi0805 — APPROVED** [review]

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java (hunk 8 above)
Comment (sbayer55) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java:20]: Good find! In one of the other PRs I moved a few classes in this package to package private. I'll make sure this class is on the list

**Review by chenqi0805 — DISMISSED** [review]

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java [hunk 9]
```diff
@@ -110,12 +118,35 @@ public GenericRegexMatchOperator regexNotEqualOperator() {
...
+        final BiPredicate<Object, Object> floatEquals = (lhs, rhs) -> ((Number)lhs).floatValue() == ((Number) rhs).floatValue();
+
+        final Map<Class<?>, Map<Class<?>, BiPredicate<Object, Object>>> equalStrategy = new HashMap<>();
+
+        final Map<Class<?>, BiPredicate<Object, Object>> intOperations = new HashMap<>();
+        intOperations.put(Integer.class, (lhs, rhs) -> (int) lhs == (int) rhs);
+        intOperations.put(Float.class, floatEquals);
+
+        final Map<Class<?>, BiPredicate<Object, Object>> floatOperations = new HashMap<>();
+        floatOperations.put(Integer.class, floatEquals);
```
Comment (cmanning09) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java:130]: Should this be `floatOperations.put(Integer.class, (lhs, rhs) -> (int) lhs == (int) rhs);`?

Comment (sbayer55) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java:130]: In this case, the `lhs` is of type `Float`. casting to an int case will cause precision loss. [used here](https://github.com/sbayer55/data-prepper/blob/d7e659ca01ab7cd2c50bd8473d27e385addee152/data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java#L134)

**Review by dlvenable — APPROVED** [review]

**Review by chenqi0805 — APPROVED** [review]

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/BinaryOperator.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package org.opensearch.dataprepper.expression;
+
+import org.antlr.v4.runtime.RuleContext;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+abstract class BinaryOperator<T> implements Operator<T> {
```
Comment (cmanning09) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/BinaryOperator.java:8]: Do we need an abstract BinaryOperator now that the GenericNegateOperator no longer extends this class? The only class extending this class is the `GenericEqualOperator`. All of this logic can just be pulled in and we can eliminate a layer of abstraction and simplify the code.

File: data-prepper-expression/build.gradle [hunk 11]
```diff
@@ -38,7 +38,7 @@ jacocoTestCoverageVerification {
     violationRules {
         rule { //in addition to core projects rule - this one checks for 100% code coverage for this project
             limit {
-                minimum = 1.0 //keep this at 100%
+                minimum = 0.98 //keep this at 100%
```
Comment (cmanning09) [inline data-prepper-expression/build.gradle:41]: Are you following up to bring this back to 100 or is this latest change permanent? If it's permanent, let remove this comment.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/BinaryOperator.java (hunk 10 above)
Comment (sbayer55) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/BinaryOperator.java:8]: After updating the GenericNotEqualOperator, the `BinaryOperator` class is unnecessary. We could remove it in a future PR or update other binary operators to extend it. The `BinaryOperator` class can remove a significant amount of repeated code. Lets choose a direction and I'll open a github issue.

File: data-prepper-expression/build.gradle (hunk 11 above)
Comment (sbayer55) [reply data-prepper-expression/build.gradle:41]: Added back in [this PR](https://github.com/opensearch-project/data-prepper/pull/1166/files#diff-0dddda0580c72f23a84a7b85ed7a151fc951a45ae09cd7d2861566213a3fe7f4R41)

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java [hunk 12]
```diff
@@ -17,7 +17,7 @@
 import java.util.function.BiPredicate;
 
 @Named
-public class OperatorFactory {
+class OperatorConfiguration {
     public final BiPredicate<Object, Object> regexEquals = (x, y) -> ((String) x).matches((String) y);
     public final BiPredicate<Object, Object> equals = Objects::equals;
```
Comment (cmanning09) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/OperatorConfiguration.java:22]: `equals` is no longer used with these changes.

**Pull request merged**